File: app/Models/TelemetryReading.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;

class TelemetryReading extends Model
{
    public const SOURCES = [
        'simulation_auto',
        'simulation_manual',
        'simulation_scenario',
        'iot',
    ];

    protected $fillable = [
        'greenhouse_id',
        'temperature',
        'soil_humidity',
        'ambient_humidity',
        'luminosity',
        'water_level',
        'irrigation_status',
        'device_status',
        'source',
        'recorded_at',
    ];

    protected function casts(): array
    {
        return [
            'temperature' => 'decimal:2',
            'soil_humidity' => 'decimal:2',
            'ambient_humidity' => 'decimal:2',
            'luminosity' => 'decimal:2',
            'water_level' => 'decimal:2',
            'recorded_at' => 'datetime',
        ];
    }

    public function greenhouse(): BelongsTo
    {
        return $this->belongsTo(Greenhouse::class);
    }

    public function alerts(): HasMany
    {
        return $this->hasMany(Alert::class);
    }

    public function isSimulated(): bool
    {
        return str_starts_with($this->source, 'simulation_');
    }
}

File: lumatek_mobile/lib/shared/models/telemetry_reading.dart
import 'json_utils.dart';

/// Lectura de sensores. Refleja App\Models\TelemetryReading.
class TelemetryReading {
  const TelemetryReading({
    required this.id,
    required this.temperature,
    required this.soilHumidity,
    required this.ambientHumidity,
    required this.luminosity,
    required this.waterLevel,
    this.irrigationStatus,
    this.deviceStatus,
    this.source,
    this.recordedAt,
  });

  final int id;
  final double temperature;
  final double soilHumidity;
  final double ambientHumidity;
  final double luminosity;
  final double waterLevel;
  final String? irrigationStatus;
  final String? deviceStatus;
  final String? source;
  final DateTime? recordedAt;

  bool get isSimulated => source?.startsWith('simulation_') ?? false;
  bool get isConnected => deviceStatus == 'connected';

  /// Acceso por clave, para recorrer las métricas junto con sus estados.
  double metric(String key) => switch (key) {
        'temperature' => temperature,
        'soil_humidity' => soilHumidity,
        'ambient_humidity' => ambientHumidity,
        'luminosity' => luminosity,
        'water_level' => waterLevel,
        _ => 0,
      };

  static String unitFor(String key) => switch (key) {
        'temperature' => '°C',
        'soil_humidity' => '%',
        'ambient_humidity' => '%',
        'luminosity' => 'lux',
        'water_level' => '%',
        _ => '',
      };

  factory TelemetryReading.fromJson(Map<String, dynamic> json) {
    return TelemetryReading(
      id: asInt(json['id']),
      temperature: asDouble(json['temperature']),
      soilHumidity: asDouble(json['soil_humidity']),
      ambientHumidity: asDouble(json['ambient_humidity']),
      luminosity: asDouble(json['luminosity']),
      waterLevel: asDouble(json['water_level']),
      irrigationStatus: asStringOrNull(json['irrigation_status']),
      deviceStatus: asStringOrNull(json['device_status']),
      source: asStringOrNull(json['source']),
      recordedAt: asDate(json['recorded_at']),
    );
  }
}
